### DISCUSSION TITLE
deploying nextjs app in docker with env vars

### DISCUSSION THREAD

**Discussion by aamorozov** [vercel/next.js#15608] [Help]
Hi! I have a use case where I build my next app in docker and deploy it to k8s cluster. There is a different set of environment variables for each environment(dev/stage/prod) so I have to rebuild the image for each environment in order to incorporate those env var sets, which is not very efficient. Is there a better way of doing that? I know building env vars into a docker image is not a best practice, but the next app needs them at build time rather than runtime, so I'm a bit at a loss. Thank you!

**Comment by eric-hc** [comment]
I'm struggling to do the same thing right now. Thought it would be best to use the native `.env.*` structure https://nextjs.org/docs/basic-features/environment-variables, but not having much luck. [Some additional discussion here](https://github.com/vercel/next.js/discussions/12077)

**Reply by eric-hc** [reply]
It would even just be nice to be able to tell Next.js what `.env` to load using `next start`. As far as I'm aware right now, we are just stuck with `next dev` loading `.env.development` and `next start` loading `.env.production`

**Reply by eric-hc** [reply]
@aamorozov [this might be a temporary solution](https://stackoverflow.com/a/61750672)

**Reply by eric-hc** [reply]
I have also been getting different runtime permission issues running `next dev` vs. `next start`, I'm on OpenShift though, but no issues running the container locally. 

Usually this `Error: EACCES: permission denied`

**Reply by aamorozov** [reply]
I think it's a slightly different issue. I use one `.env` file without prefixes. I write the env vars to it from SSM, but then those vars get built-in to the image, which is not great and I have to rebuild almost the entire image anytime I promote dev -> stage, stage -> prod.

**Reply by revanth369** [reply]
Hi @aamorozov, got any solution? I am in the same situation from promoting the builds from QA->stage and stage->PROD without re-building.

**Reply by aamorozov** [reply]
Unfortunately not, unless you're compiling next for `serverless` target, using .env file and rebuilding docker images for each environment seems like the only way to go. You can still create a base image with `node_modules` for example, to speed up the build, but the part that actually builds the app will need to be re-run in each environment.

**Comment by Mozart409** [comment]
This is still relevant. Is there any help?
